### DISCUSSION TITLE
osu!mania : Extra random mod algorithm (like s-random/h-random)

### DISCUSSION THREAD

**Discussion by engel-arthur** [ppy/osu#27471] [Ideas]
Hello.

I couldn't find anything about this after searching here (I tried random mod/s-random/random algorithm...) and on google (apart from very old forum threads), so I'm making a new discussion, sorry if a decision has already been made on this before.

As of right now, random mod in osu!mania switches around the lane order, which is great for training, but I wondered if it would be possible to implement a personalization for the mod, or an extra mod, that would shuffle each individual note (in the same way as s-random in games like beatoraja and lr2), instead of transposing every note from one lane to another.

It could optionally shuffle consecutive notes to different lanes to avoid jacks (which would be similar to h-random).

You guys already implemented the invert mod, which is super great to train ln, and s-random would be cool to train weird and unusual patterns, as well as increasing beatmap replayability (though this is subjective of course, I know some people hate random but I think the chaotic aspect of it makes it fun).

This mod should ideally not shuffle notes on the special lane (the +1 on nk+1 charts)

I'm a bit rusty when it comes to C#, but I can try to whip up a prototype if needed, if you think the idea is ok but you don't have time in your roadmap.

Have a nice evening/day!

**Comment by Rekunan** [comment]
Can you provide of video of what this would look like? I'm having trouble understanding just from reading.

**Comment by engel-arthur** [comment]
**EDIT** : _Here is an explanation that might be better than mine, you can clearly see the difference between random (like in osu), and s-random (my proposal):_
https://iidx.org/compendium/random#s-random

Sorry if I was unclear (if you have suggestions to improve the clarity of my original post, feel free, I removed some clutter already to make my point more concise), here are two videos, one with an unaltered chart, and one with an s-randomed (s-random being the name I'll use to refer to my proposal) version:

End Time - Another
https://www.youtube.com/watch?v=TW-qir8yJb4 (regular)
https://www.youtube.com/watch?v=rriPcc-xGss (S-random)

Notice how in the start, in the regular version, there are two consecutives notes on the same lane (second and third notes of the chart), and those same notes in the s-randomed version are on different lanes. This would not happen with the current random algorithm, both notes would still be on the same lane, just a different one than the original. You might notice in the video that this may lead to a lot of jacks appearing, hence why a variant, called h-random, exists, which makes it so consecutive notes are not randomized to the same lane (though I think s-random is fun as it is).

To use another example, take a simple jack like this before randomization:
![regular_jack](https://github.com/ppy/osu/assets/44979681/05c5fc4b-2cee-46a1-b98b-d5c8403d8f8e)

With the current random mod, this jack will still be a jack, just switched to a random lane, and subsequent notes which come from the same lane in the original will always be on this lane. It could look like this for example:
![regular_random_jack](https://github.com/ppy/osu/assets/44979681/bab193dd-5ed5-4a43-b5fc-0607604c5155)

With s-random, the jack would potentially become another pattern entirely, even a stair (and vice-versa, stairs might become jacks or trills, or anything, since every note is random). here's an example of what this jack could turn into:
![s_randomed_jack](https://github.com/ppy/osu/assets/44979681/dd3c83f3-9525-4fc4-ad21-50b66af4987c)



If I had to provide a (very) naïve pseudo code algorithm, it would look like this (of course this wouldn't work with chords or lns but it's to give an idea):
```
for note in beatmap.notes:
  var new_lane = random_int_between(0, MAX_LANES)
  note.lane = new_lane // each individual note is switched to a random lane, instead of whole lanes of notes getting switched.
 ```
 
 TL;DR: Every note is still present (ln or rice), with the same timings as the original beatmap, but they get shuffled to other lanes.

**Reply by Rekunan** [reply]
Ok yeah this makes much more sense, ty.
